### PR TITLE
Dropping CMAKE_PREFIX_PATH in scripts

### LINKED ISSUE #29: Errors are not caught because set -e does not apply to functions
Many errors during the build are not caught because `set -e` (in [do_everything.so](https://github.com/Intermodalics/ros_android/blob/kinetic/do_everything.sh#L4) and other shell scripts) does not apply to bash functions. If a command within a function fails, it simply continues. Only the result of the last command is returned as the function's return value in absence of an explicit `return` statement, which might trigger an error in the caller if `set -e` was effective there.

One example that can never fail is [apply_patch](https://github.com/Intermodalics/ros_android/blob/kinetic/utils.sh#L61-L69):
```bash
# Check if patch hasn't already applied and apply it
apply_patch() {
    echo "Checking patch: $1"
    if patch -p0 -N --dry-run --silent -d $prefix < $1;
    then
        patch -p0 -N -d $prefix < $1
    fi
    echo ''
}

apply_patch foo.patch   # always succeeds without error even if foo.patch does not exist or not apply
```

The solution would be to repeat `set -e` in every function or add ` || return $?` to all commands that are likely to fail.

### PR DESCRIPTION
**Pull request by jubeira**
This PR is a followup on #66.

As explained here https://github.com/Intermodalics/ros_android/pull/66#issuecomment-448754371 `CMAKE_PREFIX_PATH` was actually not being used; `CMAKE_FIND_ROOT_PATH` is the one that counts. Therefore, I think it's best not to use it in the scripts and let `catkin` handle it.

I also corrected some paths that were wrong in `build_cpp` (from #69, the path passed as parameter to `build_cpp` was `$prefix`, not `$prefix/target`, so the boost defines were malformed). In any case they don't seem to make any difference.

Finally, in the end the problem described in https://github.com/Intermodalics/ros_android/pull/66#issuecomment-446229057 was because the patch to `libtiff` was breaking `OpenCV` patch (at least the last two chunks; in particular the one before the last one was copying the missing file, and therefore the error). This PR also fixes those patches, and verifies that each patch is applied correctly. Unfortunately, `patch` command always returns `0`: when the patch is skipped, when it's applied correctly, but also when it's malformed or when it fails. That's why I'm just parsing the output in bash. Closes #29.

I've built everything from scratch and I didn't have any issues.

### CODE REVIEW

File: do_everything.sh [hunk 1]
```diff
@@ -206,6 +206,9 @@ if [[ $skip -ne 1 ]] ; then
     # Patch opencv - Fix installation path
     apply_patch $my_loc/patches/opencv.patch
 
+    # Patch libtiff - Fix missing include
+    apply_patch $my_loc/patches/libtiff.patch
```
Comment (meyerj) [inline do_everything.sh]: Why did you separate this patch? It is still applied to the 3rd-party `libtiff` inside the `opencv3` package. This version cannot be used directly, it is only for internal use from within OpenCV.

File: utils.sh [hunk 2]
```diff
@@ -62,11 +62,22 @@ apply_patch() {
...
+    result=`cat /tmp/last_patch_output`
+    if [[ $result =~ .*FAILED.* ]]; then
+        die "Patch failed. Error message: $result"
+    elif [[ $result =~ .*malformed.* ]]; then
+        die "Malformed patch. Error message: $result"
+    elif [[ $result =~ .*The\ text\ leading.* ]]; then
+        die "Wrong target file for patch. Error message: $result"
+    fi
+
+    echo "$result"
```
Comment (meyerj) [inline utils.sh]: Would it be sufficient to abort on error? `patch` should return a non-zero exit status on failure. Adding `set -e` inside `apply_patch()` would be sufficient.

For capturing the output
```
result=$(patch -p0 -N "$@" < $patch 2>&1)
```
would be an option without the temporary file. The output of the `patch --dry-run --silent` invocation can probably be ignored unconditionally.

**Comment by ivanpauno** [comment]
LGTM

I think the patch command returns error both on skipping (already applied or reversed) and failure, and returns 0 when correct. But that is still problematic. Maybe, we could replace this trick https://github.com/Intermodalics/ros_android/blob/94328ce5e70b88328ab217732cc66238fa2a118b/utils.sh#L67-L79 with using the reverse patch option (-R):
- Applying it with dry run. If ok apply the patch.
- If not, apply the reversed patch with dry-run option. If already applied should return 0, and if not we can trigger an error.
I think it is a little clearer than parsing the output for the different errors.

File: patches/opencv.patch [hunk 3]
```diff
@@ -101,13 +101,3 @@
 +            DESTINATION ${OPENCV_OTHER_INSTALL_PATH} COMPONENT libs)
    endif()
  endif()
---- catkin_ws/src/opencv3/3rdparty/libtiff/CMakeLists.txt
```
Comment (ivanpauno) [inline patches/opencv.patch]: I think the problem was that a new line is needed here.

**Review by ivanpauno — APPROVED** [review]
LGTM

I think the patch command returns error both on skipping (already applied or reversed) and failure, and returns 0 when correct. But that is still problematic. Maybe, we could replace this trick https://github.com/Intermodalics/ros_android/blob/94328ce5e70b88328ab217732cc66238fa2a118b/utils.sh#L67-L79 with using the reverse patch option (-R):
- Applying it with dry run. If ok apply the patch.
- If not, apply the reversed patch with dry-run option. If already applied should return 0, and if not we can trigger an error.
I think it is a little clearer than parsing the output for the different errors.

File: patches/opencv.patch (hunk 3 above)
Comment (jubeira) [reply patches/opencv.patch]: yes; in any case I'd rather have it separated just in case.

Comment (meyerj) [reply patches/opencv.patch]: Indeed, a newline was missing as expected by the previous patch. Apparently I edited by hand. Sorry for that.

Additionally I had to remove a trailing space in the last line (`# Find libm, if available`) to apply the patch without "fuzz". I would prefer to reintegrate the patch in `opencv.patch` (see [comment here](https://github.com/Intermodalics/ros_android/pull/74#discussion_r243388437)).

File: utils.sh (hunk 2 above)
Comment (ivanpauno) [reply utils.sh]: Only doing set -e will also trigger errors when patch is already applied (error and already applied patch both return 1). I proposed a fix in: https://github.com/Intermodalics/ros_android/pull/75

Comment (meyerj) [reply utils.sh]: > Only doing set -e will also trigger errors when patch is already applied (error and already applied patch both return 1). I proposed a fix in: #75

Not if the failing command is the condition of an `if` statement. See also [my comment](https://github.com/Intermodalics/ros_android/pull/75/files#r243579751) in #75.

File: do_everything.sh (hunk 1 above)
Comment (jubeira) [reply do_everything.sh]: Alright, let's do a single patch as you propose.

**Comment by jubeira** [comment]
Thanks for the comments @meyerj @ivanpauno; I'm running a test from scratch with all of this. I'll merge the changes if everything goes OK.

**Comment by jubeira** [comment]
Test ran properly; I'm merging now.

**Pull request merged**